### PR TITLE
Implementing tag features.

### LINKED ISSUE #6: As a user, I want to tag the problems
... so that I can better organize my problems for future retrieval.

### LINKED ISSUE #10: As a user, I want to tag problems based on question types (e.g. graph questions, DP, etc)
... so that I can practice different kinds of problems.

_User story for epic in issue #6_

### LINKED ISSUE #11: As a user, I want to tag problems that I have failed
...  so that I can review them again in the future.

_User story for epic in issue #6_

### LINKED ISSUE #12: As a user, I want to create custom tags
... so I can organize the problems as I see fit

_User story for epic in issue #6_

### LINKED ISSUE #20: As a user, I want to tag problems based on difficulty level
... so I can keep track of my progress and strength in various categories.

### LINKED ISSUE #50: Implementing add, list, edit and delete tag features
- [x] 'newtag' to add new tag

- [x] 'lstag' to list all tags

- [x] 'deletetag' to delete tag

- [x] 'edittag' to edit tag

### PR DESCRIPTION
**Pull request by LuWenQ**
Implementing add, list, edit and delete tag features without UI, storage, and tests.
Closes #6, closes #10, closes #12, closes #20, closes #50 and closes #11

### CODE REVIEW

File: src/main/java/seedu/algobase/commons/core/Messages.java [hunk 1]
```diff
@@ -7,7 +7,8 @@
 
     public static final String MESSAGE_UNKNOWN_COMMAND = "Unknown command";
     public static final String MESSAGE_INVALID_COMMAND_FORMAT = "Invalid command format! \n%1$s";
-    public static final String MESSAGE_INVALID_PERSON_DISPLAYED_INDEX = "The Problem index provided is invalid";
-    public static final String MESSAGE_PERSONS_LISTED_OVERVIEW = "%1$d problems listed!";
-
+    public static final String MESSAGE_INVALID_PROBLEM_DISPLAYED_INDEX = "The Problem index provided is invalid";
+    public static final String MESSAGE_PROBLEMS_LISTED_OVERVIEW = "%1$d problems listed!";
+    public static final String MESSAGE_INVALID_TAG_FORMAT = "Invalid tag fromat! \n";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/commons/core/Messages.java]: typo: format

File: src/main/java/seedu/algobase/logic/commands/AddCommand.java [hunk 2]
```diff
@@ -37,7 +37,7 @@
             + PREFIX_SOURCE + "Kattis";
 
     public static final String MESSAGE_SUCCESS = "New Problem added: %1$s";
-    public static final String MESSAGE_DUPLICATE_PERSON = "This Problem already exists in the algobase";
+    public static final String MESSAGE_DUPLICATE_PROBLEM = "This Problem already exists in the algobase";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/AddCommand.java]: Is your fork too far behind upstream/master? Try to rebase from the upstream before starting working next time.

File: src/main/java/seedu/algobase/logic/commands/AddTagCommand.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import seedu.algobase.logic.commands.exceptions.CommandException;
+import seedu.algobase.model.Model;
+import seedu.algobase.model.tag.Tag;
+
+/**
+ * Adds a Tag to the algobase.
+ */
+public class AddTagCommand extends Command {
+
+    public static final String COMMAND_WORD = "newtag";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/AddTagCommand.java]: Can we have `addtag` so that it's consistent with `add` and `addplan`?

File: src/main/java/seedu/algobase/logic/commands/AddTagCommand.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a Tag to the algobase. "
+            + "Parameters: "
+            + PREFIX_TAG + "TAG NAME";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/AddTagCommand.java]: No example is provided in the usage.

File: src/main/java/seedu/algobase/logic/commands/AddTagCommand.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        requireNonNull(tag);
+        toAdd = tag;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
+        if (model.hasTag(toAdd)) {
+            throw new CommandException(MESSAGE_DUPLICATE_PROBLEM);
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/AddTagCommand.java]: Why is it `DUPLICATE_PROBLEM` here?

File: src/main/java/seedu/algobase/logic/commands/EditTagCommand.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    public static final String COMMAND_WORD = "edittag";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the details of the Tag identified "
+            + "by the index number used in the displayed Tag list. "
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_TAG + "TAG] "
+            + "Example: " + COMMAND_WORD + " 1 " + PREFIX_TAG + "Easy";
+    public static final String MESSAGE_EDIT_PROBLEM_SUCCESS = "Edited Tag: %1$s";
+    public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/EditTagCommand.java]: Currently, Tag only has one field. (Probably you copied this line from `EditCommand`...)

File: src/main/java/seedu/algobase/logic/commands/EditTagCommand.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    public static final String MESSAGE_DUPLICATE_TAG = "This Tag already exists in the algobase.";
+
+    private final Index index;
+    private final Optional<String> name;
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/EditTagCommand.java]: Can you justify the usage of `Optional` here? In `EditCommand` it is used because not all fields need to be provided, however here `name` is required right?

File: src/main/java/seedu/algobase/logic/commands/ListTagCommand.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import static seedu.algobase.model.Model.PREDICATE_SHOW_ALL_TAGS;
+
+import seedu.algobase.model.Model;
+
+/**
+ * Lists all tags in the algobase to the user.
+ */
+public class ListTagCommand extends Command {
+
+    public static final String COMMAND_WORD = "lstag";
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/ListTagCommand.java]: Similarly, can we have `listtag` instead?

File: src/main/java/seedu/algobase/model/Model.java [hunk 9]
```diff
@@ -84,4 +90,57 @@
...
+    void deleteTags(Tag target);
+
+    /**
+     * Adds the given Tag.
+     * {@code Tag} must not already exist in the algobase.
+     */
+    void addTag(Tag tag);
+
+    /**
+     * Adds the given Tag lists.
```
Comment (le0tan) [inline src/main/java/seedu/algobase/model/Model.java]: typo: "list"

File: src/test/java/seedu/algobase/logic/commands/DeleteTagCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package seedu.algobase.logic.commands;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static seedu.algobase.logic.commands.DeleteTagCommand.COMMAND_WORD;
+
+import org.junit.jupiter.api.Test;
+
+public class DeleteTagCommandTest {
+    @Test
+    public void testcommandword() {
```
Comment (le0tan) [inline src/test/java/seedu/algobase/logic/commands/DeleteTagCommandTest.java:10]: Naming of your tests are not consistent with AddressBook, you may refactor them in v1.3.

**Review by le0tan — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/algobase/logic/commands/AddTagCommand.java (hunk 3 above)
Comment (LuWenQ) [reply src/main/java/seedu/algobase/logic/commands/AddTagCommand.java]: I've followed the UG and I also think `addtag` is better.
I think we can change the command in UG.

Comment (le0tan) [reply src/main/java/seedu/algobase/logic/commands/AddTagCommand.java]: Sure, go ahead and change the UG.

**Review by le0tan — APPROVED** [review]

**Comment by jiayushe** [comment]
Please fix the tests so we can merge it in.

You can try rebasing them to clean up the commit history as well.

File: src/main/java/seedu/algobase/logic/parser/AddTagCommandParser.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+     * Parses the given {@code String} of arguments in the context of the AddTagCommand
+     * and returns an AddTagCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public AddTagCommand parse(String args) throws ParseException {
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_TAG);
+        if (!arePrefixesPresent(argMultimap, PREFIX_TAG)
+                || !argMultimap.getPreamble().isEmpty()) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, AddCommand.MESSAGE_USAGE));
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/parser/AddTagCommandParser.java:28]: `AddCommand.MESSAGE_USAGE` should be `AddTagCommand.MESSAGE_USAGE`.

File: src/main/java/seedu/algobase/model/AlgoBase.java [hunk 12]
```diff
@@ -95,6 +103,43 @@ public void removeProblem(Problem key) {
         problems.remove(key);
     }
 
+    //// tag methods
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/AlgoBase.java:106]: Follow the same style for separators.

File: src/main/java/seedu/algobase/model/AlgoBase.java [hunk 13]
```diff
@@ -133,6 +178,12 @@ public void setPlan(Plan target, Plan editedPlan) {
      * Removes {@code key} from this {@code AlgoBase}.
      * {@code key} must exist in the algobase.
      */
+    public void removeTag(Tag key) {
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/AlgoBase.java:181]: Group them properly.

File: src/main/java/seedu/algobase/model/AlgoBase.java [hunk 14]
```diff
@@ -146,10 +197,15 @@ public void removePlan(Plan key) {
 
     @Override
     public String toString() {
-        return problems.asUnmodifiableObservableList().size() + " problems";
+        return tags.asUnmodifiableObservableList().size() + " tags";
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/AlgoBase.java:200]: Why this change?

File: src/main/java/seedu/algobase/model/Model.java [hunk 15]
```diff
@@ -64,6 +70,7 @@
 
     /** Returns the AlgoBase */
     ReadOnlyAlgoBase getAlgoBase();
+    /////
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/Model.java:73]: Remove this.

File: docs/tutorials/AddRemark.adoc [hunk 16]
```diff
@@ -36,7 +36,7 @@ We accomplish that by returning a `CommandResult` with an accompanying message.
 .RemarkCommand.java
 [source, java]
 ----
-package seedu.address.logic.commands;
+package seedu.address.seedu.algobase.commands;
```
Comment (jiayushe) [inline docs/tutorials/AddRemark.adoc:39]: Why?

**Review by jiayushe — COMMENTED** [review]
Some comments for next PR.

**Comment by jiayushe** [comment]
Resolve most issues in #75 .
